resolwe/flow/execution_engines/workflow/__init__.py

```diff
@@ -0,0 +1,106 @@
+"""An execution engine that supports workflow specifications."""
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import collections
+
+from six import string_types
+
+from django.db import transaction
+
+from resolwe.flow.execution_engines.base import BaseExecutionEngine
+from resolwe.flow.execution_engines.exceptions import ExecutionError
+from resolwe.flow.expression_engines import EvaluationError
+from resolwe.flow.models import Data, Process
+
+
+class ExecutionEngine(BaseExecutionEngine):
+    """An execution engine that supports workflow specifications."""
+
+    name = 'workflow'
+
+    def get_output_schema(self, process):
+        """Return any additional output schema for the process."""
+        return [
+            {'name': 'steps', 'type': 'list:data:'},
+        ]
+
+    @transaction.atomic
+    def evaluate(self, data):
+        """Evaluate the code needed to compute a given Data object."""
+        expression_engine = data.process.requirements.get('expression-engine', None)
+        if expression_engine is not None:
+            expression_engine = self.get_expression_engine(expression_engine)
+
+        # Parse steps.
+        steps = data.process.run.get('program', None)
+        if steps is None:
+            return
+
+        if not isinstance(steps, list):
+            raise ExecutionError('Workflow program must be a list of steps.')
+
+        # Expression engine evaluation context.
+        context = {
+            'inputs': {},
+            'steps': collections.OrderedDict(),
+        }
+
+        for index, step in enumerate(steps):
+            try:
+                step_id = step['id']
+                step_slug = step['run']
+            except KeyError as error:
+                raise ExecutionError('Incorrect definition of step "{}", missing property "{}".'.format(
+                    step.get('id', index), error
+                ))
+
+            # Fetch target process.
+            process = Process.objects.filter(slug=step_slug).order_by('-version').first()
+            if not process:
+                raise ExecutionError('Incorrect definition of step "{}", invalid process "{}".'.format(
+                    step_id, step_slug
+                ))
+
+            # Process all input variables.
+            step_input = step.get('input', {})
+            if not isinstance(step_input, dict):
+                raise ExecutionError('Incorrect definition of step "{}", input must be a dictionary.'.format(
+                    step_id
+                ))
+
+            data_input = {}
+            for input_name, input_value in step_input.items():
+                if isinstance(input_value, string_types):
```

Comment:
Some input values can be in lower levels of the dict, _e.g._ in `alignment-star` part of the input is given as:
```
{
    'filtering': {
        'outFilterType': 'Normal',
        'outFilterMultimapNmax': 2,
        'outFilterMismatchNmax': 10
    }
    ...
} 
```